ci: add M1 Macs to matrix

alaviss opened the pull request:
## Summary
GitHub now provides free M1 machines for CI, which means we get to have
M1 test coverage and binaries!

For more information, see
https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

## Details
* Updates to download/upload compiler actions:
  * Added awareness for runner architecture
  * Added a small workaround for  `shell: python`  not being compatible
with M1 CI
* Removed  `-lm`  compiler flags for macOS, as it is not required and
was causing
  warnings that broke tests.
* `temit.nim` is modified to include `stdio.h` due to `printf` usage.
* Disabled float-to-uint tests in  `tnumeric_conversions`  for non-x86
targets due
  to reliance on undefined behavior. See
https://github.com/nim-works/nimskull/issues/1155.

alaviss committed:
ci: add M1 Macs to matrix

This required a small modification to download/upload compiler to accomodate
multiple architectures.
Refs https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

alaviss committed:
actions: deal with shell-python not available on M1

alaviss committed:
fenv, math: don't specify -lm on macOS

Xcode 15, which is used by newer macOS now warns if the flags are redundant,
breaking tests. Suppressing such warnings does not seem to be feasible, as it
was newly added and suppression flags would not work on older Xcode.

However, it appears that macOS doesn't have libm, only a stub for POSIX compliance,
so we could just omit the flag altogether.

alaviss committed:
temit: include stdio.h

Not all systems happens to have stdio.h included indirectly via nimskull
stdlib. As such, make sure that we import stdio.h if we want to test against
it.

saem commented:
Noice!

For review, Imma just keep casually keeping up the changes here, LMK when you want me to dig in.

alaviss committed:
tnumeric_conversions: disable float-to-int tests on non-x86

Currently this test relies on x86-specific behavior for floating point
conversion and will not pass on other architectures.

Refs https://github.com/nim-works/nimskull/issues/1155

saem commented on tests/misc/tnumeric_conversions.nim:
```diff
@@ -4,23 +4,27 @@ discard """
-  doAssert uint32(f) == uint32(int32(f))
-  doAssert uint16(f) == uint16(int16(f))
-  doAssert uint8(f) == uint8(int8(f))
+when defined(amd64) or defined(i386):
```
```suggestion
when defined(amd64) or defined(i386) or defined(js) or defined(vm):
```

pretty sure those should still work, barring the known js issue.

alaviss committed:
tnumeric_conversions: enable float tests for vm and js

These targets have custom conversion logic that are less reliant on hardware woes.

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina reviewed (approved):
The changes look good to me. Very cool that we now have test coverage for ARM CPUs!

@alaviss: For your info, I've made a minor change to the PR description: import -> include.

saem commented:
/merge

Awesome
